scripts/get_github_dl_url.py

```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env python3
+"""
+Extract the download URL for software hosted on github, taking into account OS and Architecture.
+"""
+import argparse
+import enum
+import platform
+import re
+import sys
+from typing import Callable, Iterable, NamedTuple
+
+import requests
+
+GITHUB_RELEASE_TEMPLATE = "https://api.github.com/repos/{}/releases/latest"
+
+OS = platform.system()
+ARCH = platform.machine()
+
+X86_64_ARCH_NAMES = {"x86_64", "arm64"}
+
+# TOOLS:
+# these are ways to test the URLs of each release asset
+# to match our operating system and architecture.
+# the special exceptions have their repo name also, for convenience.
+
+
+class StandardTool:
+    "The URL pattern for most tools we use."
+
+    TAR_GZ_PATTERN = re.compile(r"https://.*\.tar\.gz")
+    KERNEL_PATTERN = re.compile(OS, re.IGNORECASE)
+
+    if OS == "Darwin":
+        # some projects ship "universal binaries" which support x86 and ARM.
+        # they usually use "all" in the "arch" portion.
+        if ARCH in X86_64_ARCH_NAMES:
+            ARCH_PATTERN = re.compile("all|" + "|".join(X86_64_ARCH_NAMES))
+        elif ARCH == "arm64":
+            ARCH_PATTERN = re.compile("all|arm64")
+        else:
+            sys.exit(f"Unsupported architecture {ARCH}")
+    elif OS == "Linux":
+        if ARCH in X86_64_ARCH_NAMES:
+            ARCH_PATTERN = re.compile("|".join(X86_64_ARCH_NAMES))
+        else:
+            sys.exit(f"Unsupported architecture {ARCH}")
+    else:
+        sys.exit(f"Unsupported OS {OS}")
+
+    PATTERNS = TAR_GZ_PATTERN, KERNEL_PATTERN, ARCH_PATTERN
+
+    @classmethod
+    def url_matches(cls, url: str) -> bool:
+        return all(pat.search(url) for pat in cls.PATTERNS)
+
+
+class Nooba:
+    "Nooba's URL pattern."
+
+    repo = "noobaa/noobaa-operator"
+    arch = "mac" if ARCH == "Darwin" else ARCH
```

Comment:
arch = "mac" if ARCH == "Darwin" else "linux"

Reply:
I just realized it should be OS too 😅